doubleChainSwap.md

```diff
@@ -0,0 +1,184 @@
+# Double Chain Swap
+
+- Author: Keith Lim
+- Created: September 12 2021
+- License: CC0
+- Difficulty: Intermediate
+- Ergo Playground Link: [Double Chain Swap Contract](https://scastie.scala-lang.org/EPnXlipFRgaGRSL8mQf0mw)
+
+## Description
+
+In this example, we explore the double-chain swap contracts which allow swapping of different tokens using ergoscript.
+
+For our example at hand, consider a scenario where FstHodler and SndHodler want to exchange their tokens.
+They agree that FstHodler will give SndHodler 60 tokens of type FST in exchange for 60 tokens of type SND.
+
+In the interactive example below, we explore how to do a test swap between FstHodler and SndHodler.
+
+## Code
+
+#### [Click Here To Run The Code Via The Ergo Playground](https://scastie.scala-lang.org/EPnXlipFRgaGRSL8mQf0mw)
+
+```scala
+import org.ergoplatform.compiler.ErgoScalaCompiler._
+import org.ergoplatform.playgroundenv.utils.ErgoScriptCompiler
+import org.ergoplatform.playground._
+import org.ergoplatform.Pay2SAddress
+import sigmastate.eval.Extensions._
+import scorex.crypto.hash.{Blake2b256}
+
+///////////////////////////////////////////////////////////////////////////////////
+// Prepare A Test Scenario //
+///////////////////////////////////////////////////////////////////////////////////
+// Create a simulated blockchain (aka "Mockchain")
+val blockchainSim = newBlockChainSimulationScenario("Double-Chain Swap Scenario")
+
+// Create a new token called "FST"
+val fstToken = blockchainSim.newToken("FST")
+val fstTokenAmount = 60L
+
+// Create a new token called "SND"
+val sndToken = blockchainSim.newToken("SND")
+val sndTokenAmount = 60L
+
+// Define a swapper fstHodler (with a wallet)
+val fstHodler = blockchainSim.newParty("FST Hodler")
+
+// Define a swapper sndHodler (with a wallet)
+val sndHodler = blockchainSim.newParty("SND Hodler")
+
+///////////////////////////////////////////////////////////////////////////////////
+// Double-chain Swap Contracts - Swap Contract                                    //
+///////////////////////////////////////////////////////////////////////////////////
+
+val swapScript = s"""
+  {
+    val defined = OUTPUTS(0).R2[Coll[(Coll[Byte], Long)]].isDefined && OUTPUTS(0).R4[Coll[Byte]].isDefined
```

Comment:
R2 is always defined (likely you wanted to check for !...isEmpty ?

Reply:
Figured that it is unnecessary then.